### DISCUSSION TITLE
Remove spree_adjustments, replace with {order,line_item,shipment}_{cancellation,discounts,taxes}

### DISCUSSION THREAD

**Discussion by mamhoff** [solidusio/solidus#4295] [New Features or Ideas]
Adjustments are one of the hardest-to-parse models in our code base because they're polymorphic in two directions. The polymorphic nature of the model/table also makes it impossible (or pretty hard) to set good constraints on the database level. 

Furthermore, adjustments have to be treated differently when they're discounts or taxes: Discounts have to applied before taxes.

The codebase is pretty full of type checks regarding adjustments:
* https://github.com/solidusio/solidus/blob/master/core/app/models/spree/adjustment.rb#L34-L38
* https://github.com/solidusio/solidus/blob/master/core/app/models/spree/adjustment.rb#L109
* https://github.com/solidusio/solidus/blob/master/core/app/models/spree/adjustment.rb#L122
* https://github.com/solidusio/solidus/blob/master/core/app/models/spree/order_updater.rb#L200

I think it would be a huge improvement for the readability of the code base. The migration would not be hard to write, even in pure SQL. 

What do y'all think?

**Comment by tvdeyen** [comment]
I support that. It would be much easier for people unfamiliar with the code base and would make code better to unterstand and to maintain.

IMO adjustments should stay solely as a manual added change to the order total. That's what we already have an interface and features for (Adjustment Reasons).

Would love to hear what others think.

**Comment by jarednorman** [comment]
It's an interesting idea, certainly. My main concern here is that this would create quite the break in compatibility.

**Reply by tvdeyen** [reply]
The idea is to 

a) Add a feature toggle that is `false` by default
b) Implement the code and act on the toggle accordingly
c) Ship a version of Solidus that introduces the system, but turned off by default 
d) Ship a major version where the new system is turned on by default

But, yes. This is a major change and lot's of work.

**Comment by waiting-for-dev** [comment]
It makes a lot of sense. As you said, it's a lot of work (for sure, more than we can foresee), but if somebody is willing to take it, we can try!

**Reply by tvdeyen** [reply]
"somebody" is already working on it https://github.com/solidusio/solidus/pull/4296 ;)
